dango/account/margin/src/core.rs

```diff
@@ -0,0 +1,164 @@
+        }
+
+        let price = querier.query_price(app_cfg.addresses.oracle, denom, None)?;
+        let value = price.value_of_unit_amount(collateral_balance)?;
+        let adjusted_value = value.checked_mul(power.clone().into_inner())?;
+
+        collaterals.insert(Coin::new(denom.clone(), collateral_balance)?)?;
+        total_collateral_value.checked_add_assign(value)?;
+        total_adjusted_collateral_value.checked_add_assign(adjusted_value)?;
+    }
+
+    // Iterate over the user's limit orders and add the order value to the
+    // total collateral value.
+    let mut limit_order_collaterals = Coins::new();
+    let mut limit_order_outputs = Coins::new();
+    let orders = querier.query_wasm_smart(app_cfg.addresses.dex, QueryOrdersByUserRequest {
+        user: account,
+        start_after: None,
+        limit: None,
+    })?;
+    for (_, res) in orders {
+        // Get asset locked in the order and the asset that would be returned
+        // if the order was filled.
+        let (deposited, returned) = match res.direction {
+            Direction::Bid => (
+                Coin::new(
+                    res.quote_denom.clone(),
+                    res.remaining.checked_mul_dec_ceil(res.price)?,
+                )?,
+                Coin::new(res.base_denom.clone(), res.remaining)?,
+            ),
+            Direction::Ask => (
+                Coin::new(res.base_denom.clone(), res.remaining)?,
+                Coin::new(
+                    res.quote_denom.clone(),
+                    res.remaining.checked_mul_dec_floor(res.price)?,
+                )?,
+            ),
+        };
+        let deposited_price =
+            querier.query_price(app_cfg.addresses.oracle, &deposited.denom, None)?;
+        let returned_price =
+            querier.query_price(app_cfg.addresses.oracle, &returned.denom, None)?;
+        let deposited_value = deposited_price.value_of_unit_amount(deposited.amount)?;
+        let returned_value = returned_price.value_of_unit_amount(returned.amount)?;
+
+        let min_value = std::cmp::min(deposited_value, returned_value);
+
+        let collateral_power =
+            app_cfg
+                .collateral_powers
+                .get(&deposited.denom)
+                .ok_or(anyhow::anyhow!(
+                    "collateral power for denom {} not found",
+                    deposited.denom
+                ))?;
+        let adjusted_value = min_value.checked_mul(collateral_power.clone().into_inner())?;
+
+        total_collateral_value.checked_add_assign(min_value)?;
+        total_adjusted_collateral_value.checked_add_assign(adjusted_value)?;
+        limit_order_collaterals.insert(deposited)?;
+        limit_order_outputs.insert(returned)?;
+    }
+
+    // Calculate the utilization rate.
+    let utilization_rate = if total_debt_value.is_zero() {
+        // The account has no debt. Utilization is zero in this case,
+        // regardless of collateral value.
+        Udec128::ZERO
```

Comment:
Irrelevant to this PR, but I think we can make an optimization: if debt value is zero, then we can already return without querying the collaterals and limit orders?

Reply:
Mm but would need some refactoring, since we also want to still be able to query full HealthResponse via QueryMsg.